[rocprofiler-sdk] Improve build with system libraries.

lubosz opened the pull request:
## Motivation

This series improves building rocprofiler-sdk / rocprofv3 with system libraries.

* It fixes a bug in the CMake build when `ROCPROFILER_BUILD_GLOG` was off
* It adds the possibility to build with the system's yaml-cpp
* It fixes the build with fmt 12.1.0, keeping it backwards compatible with the currently submoduled fmt version
* It proposes updating the fmt submodule to 12.1.0

This series is the result of packaging `rocprofv3` in Arch Linux.
See https://gitlab.archlinux.org/archlinux/packaging/packages/rocprofiler/-/merge_requests/3

## Technical Details

### Add option to use system yaml-cpp.
    
Adds `ROCPROFILER_BUILD_YAML_CPP` CMake variable, defaulting to ON, which keeps the previous behaviour as default.

This enables building with a system yaml-cpp.

Since the submodule uses yaml-cpp's master branch, set the required version to the latest release 0.8.0.

### Inlcude CPack before calling it

This resolves a bug in the CMake code, which relied on the glog submodule to include CPack, which was included too late.

Resolves following CMake error when building with `-D ROCPROFILER_BUILD_GLOG=OFF`:
```
CMake Error at cmake/rocprofiler_config_packaging.cmake:124 (cpack_add_component_group):
  Unknown CMake command "cpack_add_component_group".
Call Stack (most recent call first):
  CMakeLists.txt:135 (include)
```

### Fix build with fmt 12.1.0.

This enables building with fmt versions 12.1.0 and higher by including
the now required <fmt/ranges.h> header for fmt::join.

This change is backwards compatible with the current submodule version
of fmt which is an old main branch, 15 commits above 10.1.0.

### Update fmt submodule to 12.1.0

This patch is optional and I can drop it if required.

This updates the fmt submodule to 12.1.0, which is now possible with the previous patch and a good opportunity to change the submodule to a release version. Currently it uses an old main branch, 15 commits above the 10.1.0 release.


## Test Plan

I am able to `rocprofiler-sdk` with and without system libraries with this patch set.
Running a CI build should ensure the validity of this patch set.

## Test Result

Works for me.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

lubosz committed:
rocprofiler-sdk/cmake: Inlcude CPack before calling it.

Resolves following error when building with
`-D ROCPROFILER_BUILD_GLOG=OFF`:
```
CMake Error at cmake/rocprofiler_config_packaging.cmake:124 (cpack_add_component_group):
  Unknown CMake command "cpack_add_component_group".
Call Stack (most recent call first):
  CMakeLists.txt:135 (include)
```

Signed-off-by: Lubosz Sarnecki <lubosz@gmail.com>

lubosz committed:
rocprofiler-sdk/cmake: Add option to use system yaml-cpp.

Add ROCPROFILER_BUILD_YAML_CPP cmake variable, defaulting to ON, which
keeps the previous behaviour as default.

This enables building with a system yaml-cpp.

Link rocprofiler-sdk-yaml-cpp with yaml-cpp::yaml-cpp in that case,
as is also done with fmt, glog, etc.

Since the submodule uses yaml-cpp's master branch, set the required
version to the latest release 0.8.0.

Signed-off-by: Lubosz Sarnecki <lubosz@gmail.com>

lubosz committed:
rocprofiler-sdk: Update fmt submodule to 12.1.0.

Signed-off-by: Lubosz Sarnecki <lubosz@gmail.com>

lubosz committed:
rocprofiler-sdk: Fix build with fmt 12.1.0.

This enables building with fmt versions 12.1.0 and higher by including
the now required <fmt/ranges.h> header for fmt::join.

This change is backwards compatible with the current submodule version
of fmt which is an old main branch, 15 commits above 10.1.0.

Signed-off-by: Lubosz Sarnecki <lubosz@gmail.com>

jrmadsen commented on projects/rocprofiler-sdk/cmake/rocprofiler_config_packaging.cmake:
```diff
@@ -134,6 +134,8 @@ endif()
 list(APPEND _DEB_PACKAGE_DEPENDS ${ROCPROFILER_CPACK_DEBIAN_PACKAGE_DEPENDS})
 list(APPEND _RPM_PACKAGE_REQUIRES ${ROCPROFILER_CPACK_RPM_PACKAGE_REQUIRES})
 
+include(CPack)
```
Please remove. This effectively renders everything below this line pointless. All the various `CPACK_*` variables need to be set before `include(CPack)`.

lubosz replied:
Thanks a lot for the review!

I have now left the `include(CPack)` at the end of the file while fixing the undefined `cpack_add_component_group` function by only including `CPackComponent` beforehand.

jrmadsen commented on projects/rocprofiler-sdk/cmake/rocprofiler_config_packaging.cmake:
```diff
@@ -248,37 +250,4 @@ set(CPACK_PACKAGE_VERSION
     "${CPACK_PACKAGE_VERSION_MAJOR}.${CPACK_PACKAGE_VERSION_MINOR}.${CPACK_PACKAGE_VERSION_PATCH}.${ROCM_VERSION_FOR_PACKAGE}"
     )
 
-# Make CPack use LLVM binutils so ELF64-AMDGPU (.hsaco) is supported
```
Please revert.

lubosz replied:
This line is not part of my diff, not sure where it came from but hopefully you won't see it anymore since I did a rebase onto the current `develop` branch.

jrmadsen commented on projects/rocprofiler-sdk/external/CMakeLists.txt:
```diff
@@ -131,20 +131,24 @@ if(NOT TARGET PTL::ptl-static)
+        rocprofiler-sdk-yaml-cpp
+        INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/external/yaml-cpp/include>)
+else()
+    find_package(yaml-cpp 0.8.0 REQUIRED)
```
```suggestion
    find_package(yaml-cpp 0.8.0 REQUIRED)
    target_link_libraries(rocprofiler-sdk-yaml-cpp
                        INTERFACE $<BUILD_INTERFACE:yaml-cpp::yaml-cpp>)    
```
We "link" to the `rocprofiler-sdk-yaml-cpp` target in rocprofiler-sdk. Even if the find package succeeds when `ROCPROFILER_BUILD_YAML_CPP=OFF`, we will fail to build without it.

lubosz replied:
Good catch.

Even though the build was succeeding for me before with `ROCPROFILER_BUILD_YAML_CPP=OFF`, I have now added the linking command that is done for all other external targets like `GLOG`, `FMT` and `GLOG`:
```
target_link_libraries(rocprofiler-sdk-yaml-cpp INTERFACE yaml-cpp::yaml-cpp)
```

jrmadsen commented on projects/rocprofiler-sdk/external/fmt:
Can you explain why we should update the submodule? What is the benefit here?

lubosz replied:
As mentioned before I consider this patch as optional since I don't want to interfere with your submodule versioning schedule, but I see the benefits as the following:

* Developers will write newly written code compatible with fmt 12
* The current fmt submodule commit ref is not a release tag. I would recommend using release tags where possible, as is the case after my change.
* Since my patch fixes the build with that version, I saw this PR as opportunity to upgrade the submodule making it unnecessary to be done at a later time.

Please let me know if I should drop the patch so you can update the submodule in your regular submodule versioning schedule.

jrmadsen replied:
If the new fmt ref is a release tag and the previous ref is not, that is good enough for me.

jrmadsen commented:
@lubosz Regarding:

> ## Inlcude CPack before calling it
> This resolves a bug in the CMake code, which relied on the glog submodule to include CPack, which was included too late.
> 
> Resolves following CMake error when building with -D ROCPROFILER_BUILD_GLOG=OFF:
> 
> ```console
> CMake Error at cmake/rocprofiler_config_packaging.cmake:124 (cpack_add_component_group):
>   Unknown CMake command "cpack_add_component_group".
> Call Stack (most recent call first):
>   CMakeLists.txt:135 (include)
> ```

I think what you actually want is `include(CPackComponent)` not `include(CPack)` before `cpack_add_component_group` because the file MUST end with `include(CPack)`.

lubosz commented:
> @lubosz Regarding:
> 
> > ## Inlcude CPack before calling it
> > This resolves a bug in the CMake code, which relied on the glog submodule to include CPack, which was included too late.
> > Resolves following CMake error when building with -D ROCPROFILER_BUILD_GLOG=OFF:
> > ```
> > CMake Error at cmake/rocprofiler_config_packaging.cmake:124 (cpack_add_component_group):
> >   Unknown CMake command "cpack_add_component_group".
> > Call Stack (most recent call first):
> >   CMakeLists.txt:135 (include)
> > ```
> 
> I think what you actually want is `include(CPackComponent)` not `include(CPack)` before `cpack_add_component_group` because the file MUST end with `include(CPack)`.

Good catch! I have updated the patch.

lubosz commented:
I rebased the branch onto current `develop`.
Some of my include fixes were already addresses by https://github.com/ROCm/rocm-systems/commit/36d9d33d90879bfbd406498011070ebd929a56b1

I updated my patch so it adds the ones still remaining.

@jrmadsen What is missing to get this merged?

lubosz commented on projects/rocprofiler-sdk/external/CMakeLists.txt:
```diff
@@ -131,20 +131,25 @@ if(NOT TARGET PTL::ptl-static)
+        rocprofiler-sdk-yaml-cpp
+        INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/external/yaml-cpp/include>)
+else()
+    find_package(yaml-cpp 0.8.0 REQUIRED)
+    target_link_libraries(rocprofiler-sdk-yaml-cpp INTERFACE yaml-cpp::yaml-cpp)
```
The availability of this target on all imaginable "distributions/toolchains" is not relevant since it's only optional. By default the submodule is used, ensuring compatibility with distributions that do not ship `yaml-cpp`.
Although for distributions that do ship `yaml-cpp` the `ROCPROFILER_BUILD_YAML_CPP=OFF` option can be used with this patch.
With that it can be assumed that distributors that choose this option also package the upstream `yaml-cpp` CMake build.


The `yaml-cpp::yaml-cpp` target  is exposed in the following CMake block:
https://github.com/jbeder/yaml-cpp/blob/master/CMakeLists.txt#L154


In addition, the `yaml-cpp` target is deprecated and will be removed starting with version `0.10.0`
https://github.com/jbeder/yaml-cpp/blob/master/yaml-cpp-config.cmake.in#L28

lubosz commented on projects/rocprofiler-sdk/external/CMakeLists.txt:
```diff
@@ -131,20 +131,25 @@ if(NOT TARGET PTL::ptl-static)
-    rocprofiler-sdk-yaml-cpp
-    INTERFACE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/external/yaml-cpp/include>)
+    target_link_libraries(rocprofiler-sdk-yaml-cpp
+                        INTERFACE $<BUILD_INTERFACE:yaml-cpp::yaml-cpp>)
```
Thanks, fixed the formatting.
